WIP: beforeinstallprompt

marcoscaceres opened the pull request:
Nowhere near ready... will hopefully finish tomorrow.

The existing install section needs a total rewrite :(

kenchris commented on index.html:
```diff
@@ -233,6 +233,59 @@
       </p>
       <section>
         <h3>
+          Installing via automated prompting
```
Installing by prompting automatically

marcoscaceres replied:
fixed

kenchris commented on index.html:
```diff
@@ -233,6 +233,59 @@
+          Installing via automated prompting
+        </h3>
+        <p>
+          User agents that support an <a>automated install prompt</a> MUST run
```
UAs may support ... in which case it MUST run...

marcoscaceres replied:
I'm not a huge fan of MAY ... MUST constructions. Though we need to make it clear that user agent MAY support automated install prompts elsewhere. It's part of the big rewrite this whole install section needs.

kenchris commented on index.html:
```diff
@@ -233,6 +233,59 @@
+          </li>
+          <li>Let <var>event</var> be a newly constructed
+          <a>BeforeInstallPromptEvent</a> event named "beforeinstallprompt",
+          which does not bubble and is cancellable.
```
, and which is

kenchris commented on index.html:
```diff
@@ -233,6 +233,59 @@
+          <li>Let <var>user-choice promise</var> be the promise corresponding
+          to the event's <code>prompt()</code> method.
+          </li>
+          <li>Fire <var>event</var> at the <code>Window</code> object.
```
on the?

marcoscaceres replied:
DOM spec uses "Fire Event X at Y", so will follow that.

kenchris commented on index.html:
```diff
@@ -233,6 +233,59 @@
+          </li>
+          <li>Fire <var>event</var> at the <code>Window</code> object.
+          </li>
+          <li>Queue a task on the Realm’s settings object’s responsible event
```
should realm be uppercase? why?

marcoscaceres replied:
Realm is basically the global object: 
https://tc39.github.io/ecma262/#sec-code-realms

kenchris commented on index.html:
```diff
@@ -233,6 +233,59 @@
+              <li>If <var>event</var>'s <a>cancel flag</a> is set, terminate
+              this algorithm.
+              </li>
+              <li>Otherwise, attempt to present the end-user with a UI that
```
why attempt? You need to do it right?

kenchris replied:
or it may?

kenchris commented on index.html:
```diff
@@ -233,6 +233,59 @@
+              <li>If the previous step succeeds, resolve <var>prompt
+              promise</var>. Otherwise, reject <var>prompt promise</var> with
+              an ""AbortError", or "OperationError", or "NotAllowedError", or
+              "UnknownError" (depending on why the operation failed).
```
This may be UA determined?

kenchris commented on index.html:
```diff
@@ -410,6 +463,94 @@ <h3 id="installation-sec">
+        <p>
+          The <dfn><code>BeforeInstallPromptEvent</code></dfn> is dispatched
+          prior to activating an <a>automated install prompt</a>, allowing a
+          developer to prevent the default action for an install prompt.
```
instead of a

marcoscaceres replied:
@kenchris, I'm not sure exactly what you are wanting to change here.

marcoscaceres commented:
Thanks for the initial comments @kenchris! Much appreciated.

mgiuca commented on index.html:
```diff
@@ -410,6 +463,94 @@ <h3 id="installation-sec">
+};
+
+dictionary BeforeInstallPromptEventInit : EventInit {
+  AppBannerPromptOutcome userChoice;
```
This is something you've added here (that isn't in Chrome), right?

TBH it's not clear to me why we want to allow web developers to construct their own `BeforeInstallPromptEvent` objects that aren't being delivered by the UA. There isn't much use to this. If we do want to allow it for completeness, we need to make sure that it's not possible to use these fabricated BIPE objects to show a prompt -- more on that later.

It's strange that we go to the effort of allowing the website to specify a `userChoice` here (to simulate a real BIPE) but we don't provide any mechanism for them to override the `prompt` method. I suppose they could just subclass `BeforeInstallPromptEvent`, but if we allow that, then they can also override `userChoice` in the subclass. So I still don't see the point of this.

But if we must let the client supply a userChoice here, why is it an `AppBannerPromptOutcome` and not a `Promise<AppBannerPromptOutcome>`? The way it is now, the client has to pass the user's actual choice (accepted or dismissed) to the constructor of the event, before the simulated prompt is even shown, and the promise gets immediately resolved during construction. Shouldn't this constructor take a promise that can be resolved at a later time?

marcoscaceres replied:
> But if we must let the client supply a userChoice here, why is it an AppBannerPromptOutcome and not a Promise<AppBannerPromptOutcome>? The way it is now, the client has to pass the user's actual choice (accepted or dismissed) to the constructor of the event, before the simulated prompt is even shown, and the promise gets immediately resolved during construction. Shouldn't this constructor take a promise that can be resolved at a later time?

If someone is synthetically constructing BIP Events, they probably know what they are doing - so I would prefer just to keep it as simple as possible: i.e., they provide the value "userChoice" that it resolves. If they provide no value, the promise never resolves. If it comes up as an issue, we can set a default value.

mgiuca commented on index.html:
```diff
@@ -410,6 +463,94 @@ <h3 id="installation-sec">
+          event</dfn>:
+        </p>
+        <ol>
+          <li>Let <var>userChoice</var> be null.
```
These meta-variable names are confusing. *userChoicePromise* is `userChoice` (the `Promise<AppBannerPromptOutcome>`), and *userChoice* is the outcome that the promise is resolved with (the `AppBannerPromptOutcome`). That's messing with my head.

How about rename *userChoice* to *promptOutcome* (the outcome the promise is resolved with), and rename *userChoicePromise* to *userChoice* (the promise).

marcoscaceres replied:
Agree. They are terrible :) Will change them.

mgiuca commented on index.html:
```diff
@@ -410,6 +463,94 @@ <h3 id="installation-sec">
+            </li>
+            <li>Let <var>p</var> be a new promise.
+            </li>
+            <li>Run the following steps in parallel:
```
Missing steps. Should refer to the above "steps before initiating an install process".

marcoscaceres replied:
Yep. Coming soon.

mgiuca commented on index.html:
```diff
@@ -233,6 +233,59 @@
+          the following steps before initiating an install process:
+        </p>
+        <ol>
+          <li>If a <a>manually</a> instantiated <a>installation process</a> for
```
Why not just make this algorithm the "steps for an automatically instantiated installation process" and drop this first early-exit step?

mgiuca commented on index.html:
```diff
@@ -410,6 +463,94 @@ <h3 id="installation-sec">
+            <li>If [[\]], return a promise rejected with an
+            "InvalidStateError".
+            </li>
+            <li>Let <var>p</var> be a new promise.
```
As discussed above, we need to make sure that the following code *fails*:

```js
var e = new BeforeInstallPromptEvent({});
e.preventDefault();
e.prompt()
    .then(() => console.log('prompt() succeeded'))
    .catch(e => console.error('prompt() failed: ' + e));
```

That is, you must not be able to fabricate a `BeforeInstallPromptEvent` object, call `preventDefault` on it (to set its cancel flag), then call `prompt` on it to show an install banner without the UA's permission.

As I said above, I'm not sure why we allow these events to be fabricated at all, but if we do, there needs to be a check that this event was created by the UA and not by the site itself.

marcoscaceres replied:
I've not checked this in yet, but it now says:

> If the event's trusted flag is not set, return a promise rejected with a "NotAllowedError", optionally warning the developer that only trusted events are allowed to perfrom this operation.

mgiuca commented on index.html:
```diff
@@ -410,6 +463,94 @@ <h3 id="installation-sec">
+            <code>userChoice</code> attribute
+          </h4>
+          <p>
+            The <dfn><code>userChoice</code> attribute</dfn>, when getting,
```
"when getting"? Can we reword this?

mgiuca commented on index.html:
```diff
@@ -410,6 +463,94 @@ <h3 id="installation-sec">
+
+[Constructor(DOMString typeArg, optional BeforeInstallPromptEventInit eventInit)]
+interface BeforeInstallPromptEvent : Event {
+    readonly attribute Promise&lt;AppBannerPromptOutcome&gt; userChoice;
```
I find the fact that this is a promise attribute (as opposed to a function that returns a promise) really weird. I know it's what Chrome does. I wonder if it's too late to change it?

It causes a lot of weirdness because there are side-effects upon reading the attribute. For example, in Chrome a user-constructed `BeforeInstallPromptEvent` logs a console error just by the event object itself being printed to the console, because it tries to read the `userChoice` attribute. And [this](https://bugs.chromium.org/p/chromium/issues/detail?id=655877).

marcoscaceres replied:
Right now, yes. Still need to clarify what that is for? Let's discuss that in the bug.

marcoscaceres replied:
Yes, it's actually really irritating :) Prompt should just return the user choice. 

This would be so much nicer:  

```
e.prompt().then(obj => obj.userChoice);
```

marcoscaceres replied:
>  I wonder if it's too late to change it?

I certainly wouldn't be opposed. I guess we need some telemetry data on usage to change it - and some coordination with existing sites that are using this. 

> I find the fact that this is a promise attribute (as opposed to a function that returns a promise) really weird. I know it's what Chrome does. I wonder if it's too late to change it?

There are lots of these in the platform, FWIW. https://github.com/heycam/webidl/issues/186#issuecomment-253526030

mgiuca commented on index.html:
```diff
@@ -410,6 +463,94 @@ <h3 id="installation-sec">
+          </h4>
+          <p>
+            The <dfn><code>userChoice</code> attribute</dfn>, when getting,
+            returns a promise.
```
returns *userChoicePromise*.

I know that variable is kind scoped to the `BeforeInstallPromptEvent` constructor... hmm. I think we need some way to refer to these promises across the different algorithms. There's also weirdness that the "steps before initiating an install process" refer to these promises as "the promise corresponding to the event's *X* method", rather than by name.

marcoscaceres replied:
Agree. This is painful. 

My proposal would be that BIP has one internal "prompt" promise. That makes the whole design much more sensical. That promise resolves after prompt() succeeds and returns some kind of `UserPromptResult` object as proposed above.

mgiuca commented on index.html:
```diff
@@ -441,6 +583,17 @@ <h3 id="installation-sec">
         </div>
         <section>
           <h4>
+            <code>onbeforeinstallprompt</code> attribute
```
Should this come after onappinstalled, to match the order they appear in the IDL?

marcoscaceres replied:
Yeah, will move it.

marcoscaceres commented on index.html:
```diff
@@ -233,6 +233,59 @@
+              </li>
+              <li>If the previous step succeeds, resolve <var>prompt
+              promise</var>. Otherwise, reject <var>prompt promise</var> with
+              an ""AbortError", or "OperationError", or "NotAllowedError", or
```
Yep, will add all the links... don't want to add them yet as will have to rewrite this a few times yet.

marcoscaceres commented on index.html:
```diff
@@ -233,6 +233,59 @@
+              <li>In parallel, await the end-user's decision.
+              </li>
+              <li>Let <var>user choice</var> be the result of the previous
+              step, represented by either the string "accepted" or "dismissed".
```
I'll fix all markup related issues after we are happy with the overall text.

mgiuca commented on implementation/BeforeInstallPromptEvent.js:
```diff
@@ -0,0 +1,143 @@
+/*globals BeforeInstallPromptEvent, DOMException*/
```
Should this file be surrounded in a closure so it hides all the other variables?

mgiuca commented on implementation/BeforeInstallPromptEvent.js:
```diff
@@ -0,0 +1,143 @@
+  });
+}
+
+window.addEventListener("beforeinstallprompt", async function(event) {
```
It's weird that this code is here. I think of this file as being a polyfill for `BeforeInstallPromptEvent`, so it provides the event object as well as the methods for simulating the system sending those events (`notifyBeforeInstallPrompt`). It shouldn't also have a handler for the event which automatically calls preventDefault and prompt.

It's even weirder because `index.html` has its own BIP handler which may race with this one.

mgiuca commented on implementation/index.html:
```diff
@@ -0,0 +1,256 @@
+  } {
+    const assertion = "Even with success, prompt() can only be called once";
+    try {
+      ev.prompt();
```
Why does this call `prompt` without first calling `preventDefault`?

I'm confused because you have another handler (in BIPE.js) that calls preventDefault and prompt. This seems to be testing that that event was handled first, before this one... but as we discussed previously the order isn't guaranteed. I think it would be best to have just a single bipe handler (in this file, not the other one) that tests all of these things.

marcoscaceres commented on implementation/BeforeInstallPromptEvent.js:
```diff
@@ -0,0 +1,143 @@
+/*globals BeforeInstallPromptEvent, DOMException*/
+"use strict";
+const internalSlots = new WeakMap();
```
heh, I'm not a fan of the reverse :) However, once we put this into it's own scope, as suggested in your first comment, it won't matter. Hopefully JS will get real private members in the future.

marcoscaceres commented on implementation/BeforeInstallPromptEvent.js:
```diff
@@ -0,0 +1,143 @@
+
+class BeforeInstallPromptEvent extends Event {
+  constructor(typeArg, eventInit) {
+    // WebIDL Guard. Wont be in spec in spec as it's all handled by WebIDL.
```
fixed

marcoscaceres commented on implementation/BeforeInstallPromptEvent.js:
```diff
@@ -0,0 +1,143 @@
+
+    internal.userChoice = new Promise((resolve) => {
+      internal.userChoiceHandlers = {
+        resolve,
```
So, before it held .resolve and .reject. But will change to your suggestion:
```
internal.userChoiceResolver = resolve;
```

marcoscaceres commented on implementation/BeforeInstallPromptEvent.js:
```diff
@@ -0,0 +1,143 @@
+  if (installProcesses.length) {
+    return;
+  }
+  const event = new BeforeInstallPromptEvent("beforeinstallprompt");
```
Yeah, I was thinking of doing that (I've just been commenting out the checks while I play with this). Obviously, trusted gets set automatically when the real even is created by the browser through the binding layer. 

Within a few weeks, we will have both Chrome and Firefox implementing this for real in C++, so let's keep the current sample implementation as is.

marcoscaceres commented on implementation/BeforeInstallPromptEvent.js:
```diff
@@ -0,0 +1,143 @@
+  }
+}
+
+function trackReadyState() {
```
Technically you are correct, but as this doesn't use a "await" keyword anywhere, it's fine to leave looking like a sync function.

marcoscaceres commented on implementation/index.html:
```diff
@@ -0,0 +1,256 @@
+  const e = new BeforeInstallPromptEvent("");
+  console.assert(e.prompt.length === 0, assertion);
+} {
+  const assertion = "calling prompt with unstrusted event results in NotAllowedError error";
```
fixed

marcoscaceres commented on implementation/index.html:
```diff
@@ -0,0 +1,256 @@
+}
+
+// install events
+window.addEventListener("install", (ev) => {
```
Heh. Got used to the old way 👍  fixed!

kenchris commented on implementation/BeforeInstallPromptEvent.js:
```diff
@@ -7,9 +7,16 @@ const AppBannerPromptOutcome = new Set([
 class BeforeInstallPromptEvent extends Event {
   constructor(typeArg, eventInit) {
-    // WebIDL Guard. Wont be in spec in spec as it's all handled by WebIDL.
+    // WebIDL Guard. Not in spec, as it's all handled by WebIDL.
```
Not mentioned in the spec,

kenchris commented on implementation/BeforeInstallPromptEvent.js:
```diff
@@ -27,41 +34,34 @@ class BeforeInstallPromptEvent extends Event {
-      const msg = "Untrusted events can't call prompt().";
-      throw new DOMException(msg, "NotAllowedError");
-    }
+    // if (this.isTrusted === false) {
```
remove?

marcoscaceres commented:
Ok, this is more or less ready to go now. Anyone want to be a hero and give the spec text one last look over? Maybe @patrickkettner for a fresh perspective (you can ignore the "implementation/" files - though feedback on those tests would be welcome).

@kenchris I think I've addressed all your feedback. 

I'm away this week at a training course and part of next week at JS Conf AU in Melbourne - but will try to check and makes any updates in the evening.

mgiuca commented:
I will have another look over it shortly (definitely today).

mgiuca commented on index.html:
```diff
@@ -406,25 +468,195 @@ <h3 id="installation-sec">
+          };
+
+          dictionary BeforeInstallPromptEventInit : EventInit {
+            AppBannerPromptOutcome userChoice;
```
There was a discussion about this on the previous review round that went unfinished:

[My previous statement](https://github.com/w3c/manifest/pull/506/files/4af07193642bf9ae5d3f002337c64b196b0db908#r84207025)

I still don't see the point of this `userChoice` constructor argument being not a promise, implying that you've already made the choice and now you only have to understand why*. It _might_ be useful if `userChoice` was a promise, allowing the person fabricating this event to allow the choice to be resolved at some future point. But even then, I see little point in actually supporting the event fabrication use case at all. This would all (speccing, implementation, testing) be a lot easier if we banned fabrication altogether, or if the Web Platform doesn't allow that, just letting it exist but not do much (and not specifically inventing features to support this useless use case).

If you can justify why someone (a web developer, not a UA implementor) would want to create a `BeforeInstallPromptEvent` with a `userChoice` -- even a promise -- then I'm all ears.

\* Bonus points if you get that reference.

mgiuca commented on index.html:
```diff
@@ -406,25 +468,195 @@ <h3 id="installation-sec">
+            <code>true</code>.
+            </li>
+            <li>
+              <a>In parallel</a>:
```
How can this be done in parallel? Step 2.2 is dependent on the results of Step 2.1.

kenchris replied:
in parallel means async in spec lang

mgiuca commented on index.html:
```diff
@@ -406,25 +468,195 @@ <h3 id="installation-sec">
+            <code>userChoice</code> attribute
+          </h4>
+          <p>
+            The <dfn>userChoice</dfn> attribute, when getting, return the
```
s/return/returns?

(Maybe? I really don't understand the grammar of this awkward sentence, but as you mentioned previously it is spec language.)

mgiuca commented on index.html:
```diff
@@ -406,25 +468,195 @@ <h3 id="installation-sec">
+            presented to the end-user.
+          </p>
+          <pre class="example" title="'beforeinstallprompt' in action">
+          const tasksThatPreventInstallation = [];
```
Can this be `[...]` to imply it's got stuff in it, rather than just being an empty array?

Another way to express this would be to show a single task rather than an array. Perhaps, `waitForUserClickOnInstall`?

```js
event.preventDefault();
await waitForUserClickOnInstall;
event.prompt();
```

mgiuca commented on index.html:
```diff
@@ -406,25 +468,195 @@ <h3 id="installation-sec">
           The following extensions to the <code>Window</code> object specify
           the <a>event handler attributes</a> on which events relating to the
-          <a>installation</a> of a web application are fired.
+          <a>installation</a> of a web application are <a>fired</a>.
```
This could be taken out of this PR (as mentioned above).

mgiuca commented on index.html:
```diff
@@ -406,25 +468,195 @@ <h3 id="installation-sec">
         <div class="example">
           <p>
-            This example shows two ways of handling the "appinstalled" event.
+            This example shows two ways of handling the "<a>appinstalled</a>"
```
This could be taken out of this PR (as mentioned above).

mgiuca commented on index.html:
```diff
@@ -439,18 +671,30 @@ <h3 id="installation-sec">
           <p>
-            <dfn for="Window">onappinstalled</dfn> is an <a>event handler IDL
-            attribute</a> for the "<a>appinstalled</a>" event type. The
+            The <dfn>onappinstalled</dfn> is an <a>event handler IDL
```
This could be taken out of this PR (as mentioned above).

mgiuca commented on index.html:
```diff
@@ -2319,10 +2563,9 @@ <h3 id="applying">
-          can hint that the image is intended to serve some special
-          <dfn>purpose</dfn> in the context of the host OS (i.e., for better
-          integration).
+          When an <a>image object</a> is used as an icon, a developer can hint
```
This could be taken out of this PR (as mentioned above).

mgiuca commented on index.html:
```diff
@@ -2893,16 +3136,39 @@ <h3 id="applying">
         <li>
-          <a href="https://dom.spec.whatwg.org/#concept-event-fire"><dfn>Fire
-          an event</dfn></a>
+          <a href=
```
This could be taken out of this PR (as mentioned above).

kenchris commented on implementation/BeforeInstallPromptEvent.js:
```diff
@@ -0,0 +1,154 @@
+   *
+   */
+  class BeforeInstallPromptEvent extends Event {
+
```
unneeded newline

kenchris commented on implementation/BeforeInstallPromptEvent.js:
```diff
@@ -0,0 +1,154 @@
+  }
+
+  function trackReadyState() {
+    return new Promise((resolve) => {
```
unneeded ()

marcoscaceres replied:
fixed

kenchris commented on implementation/BeforeInstallPromptEvent.js:
```diff
@@ -0,0 +1,154 @@
+    });
+  }
+
+  async function showInstallPrompt(button) {
```
the functon calling this called the argument element, why not just call it button there as well

kenchris commented on implementation/BeforeInstallPromptEvent.js:
```diff
@@ -0,0 +1,154 @@
+    const p = new Promise((resolve) => {
+      add.addEventListener("click", () => {
+        resolve("accepted");
+        //emulate installation to home screen
```
// Emulate...

kenchris commented on implementation/BeforeInstallPromptEvent.js:
```diff
@@ -0,0 +1,154 @@
+    return p;
+  }
+
+  if(!window.BeforeInstallPromptEvent) {
```
I prefer space after if, as it is not really a function

kenchris commented on implementation/index.html:
```diff
@@ -0,0 +1,296 @@
+}
+</style>
+<button onclick="notifyBeforeInstallPrompt(this)">
+  Simulate install BIP
```
Why not write out BIP here

kenchris commented on implementation/index.html:
```diff
@@ -0,0 +1,296 @@
+  Simulate install BIP
+</button>
+<button onclick="showInstallPrompt(this)">
+  Simulate manual "Add to Home Screen"
```
Maybe add a bit of explanation to what that means

kenchris commented on implementation/index.html:
```diff
@@ -0,0 +1,296 @@
+<script>
+/*globals BeforeInstallPromptEvent, DOMException*/
+"use strict";
+//
```
why this empty comment?

kenchris commented on implementation/index.html:
```diff
@@ -0,0 +1,296 @@
+"use strict";
+//
+{
+  const assertion = "No arguments should throw";
```
no argument much throw? why would an argument throw? :-)

Or do you mean some thing like "Calling with no arguments should throw"?

kenchris commented on implementation/index.html:
```diff
@@ -0,0 +1,296 @@
+    console.assert(true, assertion);
+  }
+} {
+  const assertion = "First argument stringifies";
```
is being stringified?

kenchris replied:
I don't think the argument will stringify anything... but it itself may be stringified

kenchris commented on index.html:
```diff
@@ -224,15 +224,82 @@
+        "present an install prompt">presents an install prompt</dfn>, the
+        end-user's choice is represented as on of the following values (either
+        "accepted" or "dismissed"). These values are represented in the API of
+        this specification via <a>AppBannerPromptOutcome</a> enum.
```
via *the*

kenchris commented on index.html:
```diff
@@ -224,15 +224,82 @@
+        </p>
+        <ol>
+          <li>Wait until the <code>Document</code> of the <a>top-level browsing
+          context</a> to be <a>completely loaded</a>.
```
*is* completely loaded

kenchris commented on index.html:
```diff
@@ -224,15 +224,82 @@
+          <li>Wait until the <code>Document</code> of the <a>top-level browsing
+          context</a> to be <a>completely loaded</a>.
+          </li>
+          <li>If there is already an <a>installation process</a> being
```
prompt instead of process here? its confusing

kenchris commented on index.html:
```diff
@@ -406,25 +468,195 @@ <h3 id="installation-sec">
+          </dt>
+          <dd>
+            A boolean, initially <code>false</code>. Represents whether the
+            event has been used to <a>present an install prompt</a> (either by
```
why is the last part within ()? I would do a comma instead

marcoscaceres commented on implementation/BeforeInstallPromptEvent.js:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    prompt() {
+      if (this.isTrusted === false) {
```
Well, it's as functional as any other user-constructed event that relies on trust. You are correct that it's crippled - but that's again by design: you can still construct these events and fire them, but it's just "prompt()" that doesn't do anything - but that will be exactly the same for the real ones when constructed by developers. In that sense, what is currently implemented is correct per-spec. 

Note that I could, theoretically, take BeforeInstallPromptEvent.js and plug it into Gecko, because we support JS-based IDL bindings. So, in such an implementation I would get a "trusted" event.

marcoscaceres commented on implementation/BeforeInstallPromptEvent.js:
```diff
@@ -0,0 +1,154 @@
+  }
+
+  async function notifyBeforeInstallPrompt(element) {
+    if (document.readyState === "complete") {
```
woops!

marcoscaceres commented on implementation/BeforeInstallPromptEvent.js:
```diff
@@ -0,0 +1,154 @@
+    }
+  }
+
+  function trackReadyState() {
```
Used waitUntilReadyStateComplete

marcoscaceres commented:
Started addressing feedback... assertions in tests are now inline with clearer fail explanations now.

kenchris commented:
So maybe we should just merge this and then do the last fixed? I guess @marcoscaceres will have to merge all the commits to remove those unrelated changes and split them into a separate PR. Or maybe do that first and rebase this patch set?

marcoscaceres commented:
So, we are discussing maybe simplifying this whole things:
https://bugs.chromium.org/p/chromium/issues/detail?id=658638#c6

marcoscaceres commented:
closing, resending in pieces...
